Fancyauth.APIUtil/FancyPluginBase.cs

```diff
@@ -0,0 +1,436 @@
+                    i => ImmutableList<ActiveListener<T>>.Empty.Add(al),
+                    // update list
+                    (i, l) => l.Add(al)
+                );
+            }
+
+            // execute all listeners asynchrounusly
+            await Task.WhenAll(listeners[uniqueId].Select(t => t.Listener(thing)));
+            // save changes
+            await thing.SaveChanges();
+        }
+
+        public sealed override Task OnUserDisconnected(IUser user)
+        {
+            var t0 = OnThingRemoved(UserListeners, user.SessionId);
+            var t1 = UserDisconnected(user);
+            return Task.WhenAll(t0, t1);
+        }
+
+        /// <summary>
+        /// Stub method to be overridden to receive OnUserDisconnected-Events.
+        /// </summary>
+        /// <param name="user">User that disconnected.</param>
+        public virtual Task UserDisconnected(IUser user)
+        {
+            return Task.FromResult<object>(null);
+        }
+
+        public override Task OnChannelDeleted(IChannel channel)
+        {
+            var t0 = OnThingRemoved(ChannelListeners, channel.ChannelId);
+            var t1 = ChannelDeleted(channel);
+            return Task.WhenAll(t0, t1);
+        }
+
+        /// <summary>
+        /// Stub method to be overridden to receive OnChannelDeleted-Events.
+        /// </summary>
+        /// <param name="channel">Deleted Channel.</param>
+        public virtual Task ChannelDeleted(IChannel channel)
+        {
+            return Task.FromResult<object>(null);
+        }
+
+        private Task OnThingRemoved<T>(ConcurrentDictionary<int, ImmutableList<ActiveListener<T>>> listeners, int uniqueId) where T : IReadModifyWriteObject
+        {
+            ImmutableList<ActiveListener<T>> dontCare;
+            if (!listeners.TryRemove(uniqueId, out dontCare))
+            {
+                throw new Exception("If this happens we are even more screwed than if one is thrown in Init()");
+            }
+            return Task.FromResult<object>(null);
+        }
+
+        public sealed override Task OnUserModified(IUser user)
+        {
+            return OnThingModified(UserListeners, user, user.SessionId);
+        }
+
+        public sealed override Task OnChannelModified(IChannel channel)
+        {
+            return OnThingModified(ChannelListeners, channel, channel.ChannelId);
+        }
+
+        private async Task OnThingModified<T>(ConcurrentDictionary<int, ImmutableList<ActiveListener<T>>> listeners, T thing, int uniqueId) where T : IReadModifyWriteObject
+        {
+            // every time a user is modified, all listening functions with changed objects need to be called
+            // get all listening functions of the user
+            await Task.WhenAll(listeners[uniqueId]
+                // check if function has recently been added (Obj == null)
+                // or if the old saved object differs from the newly calculated (and directly saved) one
```

Comment:
These comments explain what the return does, so you might wanna move them there.

(and then kill the linebreak before the .Where ^^)